project.php

```diff
@@ -548,7 +548,7 @@ function do_project_info_table()
             if ($topic_id == "") {
                 $blurb = html_safe(_("Start a discussion about this project"));
                 $url = "$code_url/tools/proofers/project_topic.php?project=$projectid";
-                echo_row_a(_("Forum"), "<a href='$url'>$blurb</a>");
+                echo_row_a(_("Forum"), "<a href='$url' onclick='this.outerHTML=\"$blurb\";'>$blurb</a>");
```

Comment:
While this may work, it feels like the wrong place to fix this. Shouldn't we have logic in project_topic that prevents double thread creation? Like, maybe some sort of [double check lock](https://en.wikipedia.org/wiki/Double-checked_locking) during creation?

Reply:
Possibly, though I would not be confident implementing something like that, for fear of unintended consequences. It feels like a big solution to solve a small problem. I don't know if anything like that is currently used in the code, but if something like that is needed, I'm happy for this PR to be closed and for someone else to tackle it at some point.

Reply:
We have logic in [Project->ensure_topic()](https://github.com/DistributedProofreaders/dproofreaders/blob/master/pinc/Project.inc#L1251) that attempts to prevent double creation. The problem is that there's a race condition between the "create the topic" code we call out to phpBB for and the "link the new topic ID to the project" code immediately following. [Note that it's compounded by the topic_id being loaded when the Project object is created in project_topic.] I think it would be hard to implement a lock around this problem outside of the DB because the two page loads operate in different PHP process contexts.

Reply:
I'm definitely ignorant on a lot of the php complexities, but some quick googling lead me to https://www.php.net/manual/en/function.flock.php.
Would that be useful at all across the php processes?

Reply:
flock is really designed to be used against a workflow that is file-centric. We'd have to introduce some temporary file to lock against in one process to check on the second one and then delete when it was done on the first one. That doesn't feel like a good approach to this DB-centric workflow.

Reply:
Thanks both for pondering it